### ISSUE TITLE
Message bubble is behind keyboard.

### ISSUE THREAD

**Issue by yarodevuci** [slackhq/SlackTextViewController#580]
Most of the time if the message has new lines it ends up partially hidden behind the keyboard. Any ideas how to fix that?

**Comment by adamsrosco** [comment]
hi @yarodevuci,

is this when the keyboard first pops up, or are you unable to scroll the message for it to appear fully in view?

**Comment by yarodevuci** [comment]
@adamsrosco That's when keyboard is up and I press send button. That's how new message appears

**Comment by yarodevuci** [comment]
@adamsrosco Any word on that?

**Comment by Smnelson13** [comment]
How were you able to make the text inside of the label have a cushion around it? my label shrinks to the size of the text and ends up looking like this.
<img width="285" alt="screen shot 2017-04-24 at 12 14 19 pm" src="https://cloud.githubusercontent.com/assets/26229880/25347179/fea1749c-28e7-11e7-90f7-ff52676ba383.png">

**Comment by adamsrosco** [comment]
@yarodevuci you have to shift your content view of the tableView so the tableView will appear to shift higher. So you would need to monitor the UIKeyboardShown and dissappear events, and shift the keyboard.

**Comment by adamsrosco** [comment]
@Smnelson13 are you using the AL chat cells to generate your chat bubbles?

**Comment by adamsrosco** [comment]
@yarodevuci Sorry, read your post again, are you using height for row at index path in your project to cater for older iOS versions? I used dynamic layout on the cells and took out the heightforrow, i found that i kept getting a issue on my last chat message wrt sizing, so the cell would always also appear behind the keyboard.

NB: when using autolayout and dynamic sizing, dont forget to set the estimatedRowHeight, then your cells shouldnt appear behind the keyboard anymore.

**Comment by yarodevuci** [comment]
@adamsrosco I use UITableViewAutomaticDimension height for row. 
Can you please specify more details how I can use dynamic layout on the cells and dynamic sizing?

**Comment by adamsrosco** [comment]
It looks like you are already using autolayout.
When you run in the simulator do you get any layout warning messages? That could indicate there's atleast one constraint that may be the cause of that.

**Comment by yarodevuci** [comment]
@adamsrosco no I don't have any errors when it happens.

**Comment by adamsrosco** [comment]
mm... I will have to check this again. 
I do have a project implementing the controller again, and can have a look at it then. Will be about 3 weeks from now though. 

out of interest, is there a need for the autocorrect/autosuggest bar to show, i've disabled this in the past so the content size adjusts correctly. I am wondering in the slack controller takes this into account... it might solve your problem in the short term.

**Comment by yarodevuci** [comment]
@adamsrosco I don't see any autocorrect bar on mine . Cool I can wait, let me know later

**Comment by yarodevuci** [comment]
@adamsrosco any updates so far?

**Comment by Smnelson13** [comment]
@adamsrosco I figured it out, my solution was to create a class that gives CGFloat4 points of cushion around the whole text. that was the easiest way I personally found.

**Comment by yarodevuci** [comment]
@Smnelson13 Hi, I am totally lost of what you just said. Can you please explain in more details ?

**Comment by Smnelson13** [comment]
@yarodevuci 
Sorry, i was on my phone and tried to summarize it. `class InsetLabel: UILabel
{
  let topInset = CGFloat(4)
  let bottomInset = CGFloat(4)
  let leftInset = CGFloat(8)
  let rightInset = CGFloat(8)

  override func drawText(in rect: CGRect)
  {
    let insets: UIEdgeInsets = UIEdgeInsets(top: topInset, left: leftInset, bottom: bottomInset, right: rightInset)
    super.drawText(in: UIEdgeInsetsInsetRect(rect, insets))
  }
  
  override public var intrinsicContentSize: CGSize
  {
    var intrinsicSuperViewContentSize = super.intrinsicContentSize
    intrinsicSuperViewContentSize.height += topInset + bottomInset
    intrinsicSuperViewContentSize.width += leftInset + rightInset
    return intrinsicSuperViewContentSize
  }
}
`

I run my text through this to get the border effect I wanted. 
UsingCGRect I created insets and inserted them around the text.

**Comment by yarodevuci** [comment]
@Smnelson13 So basically u hard coded that, In my case I am using custom UITableViewCell xib

**Comment by Smnelson13** [comment]
@yarodevuci 
I used a xib as well, when I used the xib I still encountered this problem so by doing what i listed above I fixed the issue.

**Comment by yarodevuci** [comment]
@Smnelson13 So basically you got rid of xib and replaced it via code based? In that case how did you use that class? U manually create a view with text and how u put it in table view?

**Comment by Smnelson13** [comment]
@yarodevuci  I made the class of the Xib insetLabel if that's what you mean.

**Comment by yarodevuci** [comment]
@Smnelson13 By any chance do u have a full example of your code?

**Comment by Smnelson13** [comment]
@yarodevuci https://github.com/Smnelson13/NotSoCleverBot.git

**Comment by yarodevuci** [comment]
@Smnelson13 Thank you!! I ll take a look

**Comment by adamsrosco** [comment]
Hi,

our work with the controller has been pushed out. I need to have a look at your code again. How are you setting up your tableview in relation to the input field.

is the table up against the input view, or is the input view over the table?

Regards,
Rosco.

**Comment by yarodevuci** [comment]
@adamsrosco , how do I check that? I am setting it up just like example does, and am using my custom cell xib for messages, that also have bottom label for delivery status.

**Comment by johndoehelloworld** [comment]
@yarodevuci I see that you were able to use dynamic heights for your custom cell. I'm trying to do the same. I try to do the following in viewDidLoad: `tableView.estimatedRowHeight = 44
        tableView.rowHeight = UITableViewAutomaticDimension`
And then this: `override func tableView(_: UITableView, heightForRowAt _: IndexPath) -> CGFloat {
        return UITableViewAutomaticDimension
    }
    override func tableView(_: UITableView, estimatedHeightForRowAt _: IndexPath) -> CGFloat {
        return 44
    }`

But this renders my label with a height only enough to view 1 line. I have no. of lines on the label set to 0, and I'm seeing no auto-layout constraint issues. I have only the trailing, leading, and top for the said label. 

Any idea what might I be missing? Help is much appreciated!

**Comment by yarodevuci** [comment]
@johndoehelloworld have u set all the constraints like leading trailing , bottom and top?

**Comment by johndoehelloworld** [comment]
@yarodevuci Yes I did..I can't for the life of me figure out why it isn't working.. Did you do anything non standard to make this work?

**Comment by yarodevuci** [comment]
@johndoehelloworld can u attach some screenshots of ur constraints and actual tableViewCell ?

**Comment by johndoehelloworld** [comment]
![screenshot 2017-08-01 23 31 30](https://user-images.githubusercontent.com/2659271/28860478-aac2da82-7711-11e7-8720-272e831055e6.jpeg)
![screenshot 2017-08-01 23 33 31](https://user-images.githubusercontent.com/2659271/28860519-d8fe2b72-7711-11e7-85ca-2044f80bdd39.jpeg)
![screenshot 2017-08-01 23 34 39](https://user-images.githubusercontent.com/2659271/28860571-0244b1f4-7712-11e7-8524-76d83de1216a.jpeg)
![screenshot 2017-08-01 23 35 50](https://user-images.githubusercontent.com/2659271/28860615-2c88fdbc-7712-11e7-98e2-e983c893b819.jpeg)

The second and third sreenshots corresponds to the message label that needs to grow vertically to fit content. Thanks for your time!

**Comment by Smnelson13** [comment]
If you would like I can send you a Xib file that I use. I'm my newer most up to date project that I'm using this my contains are set to be greater than or = to something. You have your label constrains set but I think it might not be able to grow due to the fact the constrains are set to a exact number. By using greater than or = to you can make a constraint grow and change as need be. Then set lines to 0.

**Comment by johndoehelloworld** [comment]
Hi @Smnelson13 – can you please share with me you XIB? Thanks for offering! Are you sure you aren't talking about >= in terms of height? My constraints are about pinning the edges of the label to the sides rather than describe it's height. But again, I might be missing something. Any help is much appreciated. Thanks again!

**Comment by johndoehelloworld** [comment]
@yarodevuci : Thoughts on my screenshots? Thanks in advance for your time!

**Comment by johndoehelloworld** [comment]
@Smnelson13 For instance, see this screenshot: I increased the height of the cell in IB, and the label grows.

![screenshot 2017-08-02 18 50 24](https://user-images.githubusercontent.com/2659271/28902403-8c7b4058-77b3-11e7-8df3-3a765072aa54.jpeg)

**Comment by Smnelson13** [comment]
What I did is put a label on the left or wright depending on the user. The label has constraints that explicitly set on top and left. The bottom and right side are >= and line set to 0. I can send the xib when I get home to my computer.
